Documentation: Create documentation around AWS Secrets Manager plugin (#52181)
grafana/grafana: docs/sources/enterprise/aws-secrets-plugin.md

---
aliases:

>   - /docs/grafana/latest/enterprise/aws-secrets-plugin/

Reviewer: Plugins that require GE and involve secret management have been moved from `/enterprise/` to sections under `/setup-grafana/`. This doc should be moved to the directory `docs/sources/setup-grafana/configure-security/configure-database-encryption/`.

To align with the other secret management docs in that directory, the filename should probably also be renamed to `encrypt-secrets-using-aws-secrets-manager.md`.

CC @chri2547 for additional feedback.

```suggestion
  - /docs/grafana/latest/setup-grafana/configure-security/configure-database-encryption/encrypt-secrets-using-aws-secrets-manager/
```

Author: moved file and updated alias

> description: 'Configuring the AWS Secrets Manager plugin'

Reviewer: ```suggestion
description: 'Learn how to use AWS Secrets Manager plugin to encrypt Grafana secrets.'
```

keywords:
  - grafana
  - secrets
  - secretsmanager
  - aws
  - backend

> title: AWS Secrets Manager Integration

Reviewer: ```suggestion
title: Encrypt database secrets using AWS Secrets Manager
```

Reviewer 2: this is kind of a long title. Any way to shorten it using `menuTitle`?

weight: 2000
---

> # AWS Secrets Manager Plugin

Reviewer 2: The Title and the H1 should be the same.

By default, Grafana stores all encrypted secrets in its own SQL database. The AWS Secrets Manager plugin is available for users who wish to store secrets in [AWS Secrets Manager](https://aws.amazon.com/secrets-manager/). In the current version of Grafana, this is limited to datasource secrets, but it will eventually include all secrets (e.g. user credentials).

There are a number of configuration options available that determine secret storage behavior in AWS Secrets Manager. We also provide the ability to migrate secrets from existing Grafana instances to AWS Secrets Manager. We provide a degree of backwards compatibility so that the plugin can be quickly uninstalled, but it has some [limitations](#limitations).

> 

Reviewer: Other similar documentation lists prerequisites for using their related plugins or services. See https://grafana.com/docs/grafana/next/setup-grafana/configure-security/configure-database-encryption/encrypt-secrets-using-aws-kms/ for an example.

Also, add a note specifying that this is available in Grafana Enterprise:

```suggestion

> **Note:** Available only in [Grafana Enterprise]({{< relref "/enterprise/" >}}).

> ## Installation

Reviewer 2: no gerund, directives only.

> You can install the plugin using Grafana CLI:

Reviewer: ```suggestion
You can install the plugin using [Grafana CLI]({{< relref "../../../cli/" >}}):
```

Relative pathing here assumes the document is moved as suggested. If it isn't, the path should be `../cli/`.

```sh
grafana-cli plugins install grafana-aws-secrets-manager
```

> After the plugin is installed, it must be explicitly enabled. You can do this by adding the following to the Grafana `custom.ini` file:

Reviewer 2: Remove passive voice...I would also format this as a task. Something like:

"After you install the plugin, you must enable it. 

1. Open the xxx file.
2. Locate the `[secrets]` section.
3. Set `use_plugin` to `true`.
4. Save your changes and restart the Grafana server. (I'm not sure if that's the case, but you get the point).

> After the plugin is installed, it must be explicitly enabled. You can do this by adding the following to the Grafana `custom.ini` file:

Author: Changed into task with some slight mods.

```ini
[secrets]
use_plugin = true
```

> ## Authentication

Reviewer 2: Remove gerund

> ## Authentication

Author: Changed to `Authenticate with AWS`

By default, the plugin will attempt to authenticate and select an AWS region using parameters read from the runtime environment. For more details, see the [AWS CLI](https://aws.amazon.com/cli/) setup documentation.

Alternatively, you can configure authentication and region selection explicitly through Grafana’s configuration. The full list of available options can be found below. For more information on configuring Grafana, see [Configuring Grafana]({{< relref "../setup-grafana/configure-grafana/" >}}).

> ## Configuring the plugin

Reviewer 2: ```suggestion
## Configure the plugin
```

> To begin, create a new configuration section in the custom.ini file called `[plugin.grafana-aws-secrets-manager]`. Place all of the following configuration settings under this section. All of following settings are optional.

Reviewer 2: I would make this a mini task as you'll see in my suggestion.

> To begin, create a new configuration section in the custom.ini file called `[plugin.grafana-aws-secrets-manager]`. Place all of the following configuration settings under this section. All of following settings are optional.

Reviewer 2: I would make line 44 a small task, as i've suggested.

| `aws_region`            | AWS instance region                                                                                                                                                | us-east-1                                | _Read from runtime environment_ |
| `kms_key_id`            | AWS Key Management System access key, used for secret encryption                                                                                                   | 82065da4-3e2b-4372-87bf-664d1e488244     | _none_                          |
| `secret_name_prefix`    | String prepended to each AWS Secret Manager secret name. This can be used to avoid secret name conflicts in large organizations running multiple Grafana instances | metrics-team                             | _none_                          |
| `secret_description`    | Description applied to every secret in AWS Secrets Manager. For bookkeeping purposes only                                                                          | Metrics team datasource                  | _none_                          |

> ## Migration

Reviewer 2: Same comment here about the gerund

> ## Migration

Reviewer 2: also, migrate what to the plugin? Are customers migrating data? oh, i see now - migrating secrets. So many "Migrate your secrets to the plugin"?

You can configure Grafana to migrate your existing secrets from Grafana to the plugin on startup. This migration is a one-time blocking operation that runs on Grafana startup, meaning Grafana will not be usable until migration is complete.

**Note:** The speed of this migration is contingent on the number of secrets and system performance, so it can potentially run slowly (e.g. two secrets per second).

Once migration to the plugin has completed, and if [backwards compatibility](#backwards-compatibility) is disabled, the plugin will be required for Grafana to start. If the plugin is uninstalled unexpectedly, or if it fails to start for any reason, Grafana will also fail to start. Because the ability to migrate from the plugin back to Grafana is not yet implemented, we highly recommend keeping backwards compatibility turned on (default).

> ### Migration to plugin

Reviewer 2: Remove gerund.

> ### Migration to plugin

Author: Changed to `Initiate the migration`

Migration from Grafana to the plugin is initiated by configuration. Add the following to the Grafana `custom.ini` file:

```ini
[secrets]
migrate_to_plugin = true
```

> Secrets will be migrated to the plugin one-by-one, then deleted from the Grafana database if backwards compatibility is disabled.

Reviewer 2: I would not include the word 'will'

> Secrets will be migrated to the plugin one-by-one, then deleted from the Grafana database if backwards compatibility is disabled.

Author: took out 'will'

## Backwards compatibility

By default, after the plugin is installed and enabled, secrets will be stored in AWS Secrets Manager **and** in a legacy location in the Grafana database. If you wish to stop to storing secrets in the legacy table while the secrets plugin is enabled, you can disable backwards compatibility by adding a [feature toggle]({{< relref "../setup-grafana/configure-grafana/#feature_toggles" >}}) called `disableSecretsCompatibility`:

```ini
[feature_toggles]
enabled = disableSecretsCompatibility
```

> While secrets will not be read from this table while the plugin is active, it gives the ability to quickly uninstall the secrets plugin without losing access to all of your secrets. Because the ability to migrate from the plugin back to Grafana is not yet implemented, we highly recommend keeping backwards compatibility turned on.

Reviewer: ```suggestion
While secrets will not be read from this legacy table while the plugin is active, backward compatibility allows you to uninstall the secrets plugin without losing access to all of your secrets.

> **Note:** We have not yet implemented migrations from the plugin back to Grafana's database. Disabling backward compatibility and uninstalling the plugin can cause Grafana to fail to start. We strongly recommend keeping backward compatibility enabled, as is the default.
```

Re-emphasize the warning against disabling backward compatibility here.

## Limitations

> If secret rotation has been activated on your AWS Secrets Manager instance, the backwards compatibility feature will not be sufficient as a fallback for plugin errors. This is because legacy secrets will not be kept up to date with AWS Secrets Manager automatically. If you are planning to implement secret rotation, we recommend waiting until migration from the plugin back to Grafana is available if there is a desire to potentially uninstall the AWS Secrets Manager plugin.

Reviewer: ```suggestion
If your AWS Secrets Manager instance uses secret rotation, the backward compatibility feature will not be a sufficient fallback for plugin errors, because legacy secrets will not be kept up to date with AWS Secrets Manager automatically. If you plan to implement secret rotation, we recommend that you avoid installing this plugin until we implement migration from the plugin back to Grafana's database.
```
